language/tools/move-package/src/resolution/dependency_graph.rs

```diff
@@ -930,64 +917,122 @@ impl DependencyGraph {
+        match self.package_table.entry(name) {
+            Entry::Vacant(entry) => {
+                entry.insert(pkg.clone());
+                Ok(None)
+            }
 
             // Seeing the same package again, pointing to the same dependency: OK, return early but
             // only if seeing a package that was not on an "override path" (was created when no
             // override was active); otherwise we need to keep inspecting the graph to make sure
             // that the overrides introduced on this path correctly dominate all "uses" of a given
             // package
-            Entry::Occupied(entry) if entry.get() == &pkg => {
+            Entry::Occupied(entry) if entry.get() == pkg => {
+                if let Some(ext_graph) = external_subgraph {
+                    // when trying to insert a package into the graph as part of merging an external
+                    // subgraph it's not enough to check for package equality as it does not capture
+                    // dependencies that may differ between the internally and externally resolved
+                    // packages.
+                    let (self_deps, ext_deps) =
+                        pkg_deps_equal(name, &self.package_graph, ext_graph);
+                    if self_deps != ext_deps {
+                        bail!(
+                            "Conflicting dependencies found for '{name}' during external resolution by '{}':\n{}{}",
+                            resolver.unwrap(), // safe because external_subgraph exists
+                            format_deps("\nExternal dependencies not found:", self_deps),
+                            format_deps("\nNew external dependencies:", ext_deps),
+                        );
+                    }
+                }
                 if entry.get().overridden_path {
-                    // check if acyclic to avoid infinite recursion
+                    // check if acyclic to avoid infinite recursion - see the
+                    // diamond_problem_dep_incorrect_override_cycle (in tests) for an example of
+                    // such situation
                     self.check_acyclic()?;
-                    self.override_verify(&pkg, name, overrides)?;
+                    // inspect the rest of the graph and report error if a problem is found
+                    self.override_verify(pkg, name, overrides)?;
                 }
-                return Ok(pkg);
+                Ok(Some(pkg.clone()))
             }
 
             // Seeing the same package again, but pointing to a different dependency: Not OK unless
             // there is an override.
             Entry::Occupied(mut entry) => {
                 if let Some(overridden_pkg) = overrides.get(&name) {
-                    // override found - use it
-                    entry.insert(overridden_pkg.clone());
-                    entry.into_mut()
+                    // override found - use it and return - its dependencies have already been
```

Comment:
It seems like we only care to check if there's an override when there's a potentially conflicting set of dependencies, but if the package table is empty we just insert the package, and don't check for overrides.  Is this because if there is an override, it will be in the package table already, or is this something that we need to change? (If the former, we should add a comment on the `Vacant` case as well.

Reply:
Good point! And your assessment is correct - the override would have already been inserted (and at the point of the override insertion, the package in question is not in the overrides set). I have put a comment on the `Vacant` case accordingly.